### DISCUSSION TITLE
OnPropertyChange event access from the ViewModel

### DISCUSSION THREAD

**Discussion by kcabral817** [MarimerLLC/csla#2661] [Questions]
How do I get the event in the ViewModel that a field was changed in the BO?  I have a collection in the businessBase class that I need to consume and redirect to the message window of the view model.  But I am unable to figure out how to monitor the OnPropertyChange event from the view model.

**Comment by rockfordlhotka** [comment]
Which ViewModel? Xaml, Blazor, etc?

**Comment by kcabral817** [comment]
xaml wpf

**Comment by rockfordlhotka** [comment]
I'm assuming CSLA 5.

There is an `OnModelChanged` method you can override. In your override you _must_ invoke the base implementation or other behavior will break.

In this override you can hook and unhook the events of the Model (business domain object), including PropertyChanged. This is from memory, so might not be quite right, but something like this:

```c#
protected override void OnModelChanged(T oldValue, T newValue)
{
  if (oldValue is INotifyPropertyChanged npc)
    npc.PropertyChanged -= ModelPropertyChanged;
  if (newValue is INotifyPropertyChanged npc)
    npc.PropertyChanged += ModelPropertyChanged;
}

private void ModelPropertyChanged(object s, PropertyChangedEventArgs e)
{
  // handle the event here
}
```

**Reply by kcabral817** [reply]
I saw this in the class and implemented it as test.  But got concerned this was too complicated to get notification of an event firing.  In any event, thank you.

**Reply by rockfordlhotka** [reply]
I'd put this code in a custom base class so you don't need to write/maintain it in every viewmodel class. That way it is abstracted and isn't so complicated.

The alternative is that you could submit a PR to add a `ModelPropertyChanged` event to `ViewModelBase` in CSLA 5 (the `v5.x` branch) and 6 (the `main` branch), to put this functionality into the core framework.

**Reply by kcabral817** [reply]
I would be interested in submitting a PR for both 5 & 6.  I need to see how to do it, as I've done this before.

**Reply by rockfordlhotka** [reply]
That is fantastic!

Here's a doc on how to use git with CSLA. Please let me know if you have any questions, happy to help!

https://github.com/MarimerLLC/csla/blob/main/docs/dev/csla-github-flow.md

**Reply by kcabral817** [reply]
Rocky,

I'm on this step:  git checkout -b 123-feature-branch marimer/<maintenance-branch-name>

I'm going to use the v5.5.1 what should the name 123-feature-branch & <maintenance-branch-name> be?

**Reply by rockfordlhotka** [reply]
Sorry @kcabral817 - took me a while to get time to reply 😔

`git checkout -b 2672-propertychange marimer/v5.x`

That'll create a new branch on your workstation named `2672-propertychange` that is based off the `marimer/v5.x` branch, which is the maintenance branch for CSLA 5.

Anywhere in the doc where `marimer/main` is used, you **must** use `marimer/v5.x`, because that's the branch you are updating.
